### PR TITLE
Allow default value on lag/lead if they are coerceable to value

### LINKED ISSUE #8307: `lag` and `lead` window functions does not accept compatible default value
### Describe the bug

The window functions `lag` and `lead` allow a default value as its third parameter.
Currently in DataFusion, this needs to be Int64 even though the value selected in the function is not of type Int64.

Postgres docs say that the default value only has to be compatible: https://www.postgresql.org/docs/current/functions-window.html

### To Reproduce

```
DataFusion CLI v33.0.0
❯ create table t1 (a timestamp);
0 rows in set. Query took 0.009 seconds.

❯ insert into t1 values (1),(2),(3);
+-------+
| count |
+-------+
| 3     |
+-------+
1 row in set. Query took 0.013 seconds.

❯ select lag(a, 1, '1970-01-01T00:00:00.000000001'::timestamp) OVER(ORDER BY a) from t1;
Internal error: Expects default value to have Int64 type.
This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker
```

### Expected behavior

Allow any default value that can be coerced into the first parameter

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by simonvandel**
## Which issue does this PR close?

Closes #8307.

## Rationale for this change

Allow other default values than Int64 on lag/lead window functions.

## What changes are included in this PR?

If a default value is specified, it is now coerced if possible.
Before this PR only Int64 was allowed.

## Are these changes tested?

Yes, added some slt tests.

## Are there any user-facing changes?

Yes, more default values are now accepted.

### CODE REVIEW

File: datafusion/physical-expr/src/window/lead_lag.rs [hunk 1]
```diff
@@ -235,11 +236,27 @@ fn get_default_value(
     default_value: Option<&ScalarValue>,
     dtype: &DataType,
 ) -> Result<ScalarValue> {
-    if let Some(value) = default_value {
-        if let ScalarValue::Int64(Some(val)) = value {
-            ScalarValue::try_from_string(val.to_string(), dtype)
+    if let Some(default_value) = default_value {
+        let default_value_type = default_value.data_type();
+        if can_coerce_from(dtype, &default_value_type)
+            || (dtype.is_integer() && default_value_type.is_integer())
```
[the remark this answers is not in the document]
Comment (jayzhan211) [reply datafusion/physical-expr/src/window/lead_lag.rs:242]: Is `comparison_binary_numeric_coercion` possible for your case?

File: datafusion/physical-expr/src/window/lead_lag.rs [hunk 2]
```diff
@@ -235,11 +236,27 @@ fn get_default_value(
...
+                    &CastOptions {
+                        safe: false,
+                        format_options: Default::default(),
+                    },
+                )?,
+                0,
+            )
         } else {
-            internal_err!("Expects default value to have Int64 type")
+            Err(DataFusionError::Internal(format!(
```
[the remark this answers is not in the document]
Comment (simonvandel) [reply datafusion/physical-expr/src/window/lead_lag.rs]: Yes, thank you!

File: datafusion/physical-expr/src/window/lead_lag.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (simonvandel) [reply datafusion/physical-expr/src/window/lead_lag.rs]: `comparison_binary_numeric_coercion` is not public, but I changed to `comparison_coercion` which seems to work fine

File: datafusion/physical-expr/src/window/lead_lag.rs [hunk 3]
```diff
@@ -238,9 +238,8 @@ fn get_default_value(
 ) -> Result<ScalarValue> {
     if let Some(default_value) = default_value {
         let default_value_type = default_value.data_type();
-        if can_coerce_from(dtype, &default_value_type)
-            || (dtype.is_integer() && default_value_type.is_integer())
-        {
+
+        if comparison_coercion(&default_value_type, dtype).is_some() {
```
Comment (jayzhan211) [inline datafusion/physical-expr/src/window/lead_lag.rs:242]: It seems that you just need to change integer, since I'm not sure having `pub fn comparison_binary_numeric_coercion` is a good idea or not. At least, we can add the comment for now.

Comment (simonvandel) [reply datafusion/physical-expr/src/window/lead_lag.rs:242]: Sorry, I didn't understand what you mean by "change integer". Could you rephrase, or provide a suggestion for a change?

Comment (jayzhan211) [reply datafusion/physical-expr/src/window/lead_lag.rs:242]: Sorry, I mean coerce, not change.

Comment (jayzhan211) [reply datafusion/physical-expr/src/window/lead_lag.rs:242]:
```rust
if can_coerce_from(dtype, &default_value_type)
            || (dtype.is_integer() && default_value_type.is_integer())
```
// We only check coercion between integer.

Comment (simonvandel) [reply datafusion/physical-expr/src/window/lead_lag.rs:242]: So just to be clear, are you suggesting that the comparison should be reverted to 
```rust
if can_coerce_from(dtype, &default_value_type)
            // We explicitely allow a lossy conversion between signed and unsigned integers
            || (dtype.is_integer() && default_value_type.is_integer())
```
?

Comment (jayzhan211) [reply datafusion/physical-expr/src/window/lead_lag.rs:242]: I was suggesting `comparison_coercion` with comment.

But, I rethink about it again, now I think  
```rust
if can_coerce_from(dtype, &default_value_type)
            || (dtype.is_integer() && default_value_type.is_integer())
```
is better for here.

btw, i dont think your comment `// We explicitely allow a lossy conversion between signed and unsigned integers` is correct 🤔

Comment (jayzhan211) [reply datafusion/physical-expr/src/window/lead_lag.rs:242]: I thought you were just trying to deal with integer type. But, I think we need to cover any compatible type 🤔 ?

If that is true, we can stick with `if comparison_coercion(&default_value_type, dtype).is_some()`

Comment (jayzhan211) [reply datafusion/physical-expr/src/window/lead_lag.rs:242]: I suggest `comparison_coercion`. I read the doc, it does not seem lag/lead should only be integer.

Comment (jayzhan211) [reply datafusion/physical-expr/src/window/lead_lag.rs:242]: cc @alamb to confirm my suggestion.

Comment (alamb) [reply datafusion/physical-expr/src/window/lead_lag.rs:242]: I think coercing during physical evaluation is too late in the process -- most other coercion happens in the analyze phase so that subsequent transformations / optimizations use the same rules.  

Would it be possible to move this coercion earlier into 

https://github.com/apache/arrow-datafusion/blob/513fd052bdbf5c7a73de544a876961f780b90a92/datafusion/optimizer/src/analyzer/type_coercion.rs#L64-L63

?

**Review by alamb — COMMENTED** [review]
Thank you for this @simonvandel  and your help @jayzhan211

**Comment by alamb** [comment]
Marking as draft as i think this is no longer waiting on feedback. Please mark it ready for review if/when it is ready for another look
